ts/Stock/Indicators/DisparityIndex/DisparityIndexIndicator.ts

```diff
@@ -0,0 +1,282 @@
+                averageType,
+                ctx.type,
+                function (indicator: Highcharts.Indicator): undefined {
+                    indicator.prototype.init.apply(ctx, args);
+                    return;
+                }
+            );
+        }
+    }
+
+    public calculateDisparityIndex(
+        close: number,
+        periodAverage: number
+    ): number {
+        return correctFloat(close - periodAverage) / periodAverage * 100;
+    }
+
+    public getValues<TLinkedSeries extends LineSeries>(
+        series: TLinkedSeries,
+        params: DisparityIndexParamsOptions
+    ): (IndicatorValuesObject<TLinkedSeries>|undefined) {
+        const averageType = params.average,
+            index = params.index,
+            period = params.period as any,
+            xVal: Array<number> = (series.xData as any),
+            yVal: Array<number>|Array<Array<number>> = (series.yData as any),
+            yValLen: number = yVal ? yVal.length : 0,
+            disparityIndexPoint: Array<Array<number>> = [],
+            xData: Array<number> = [],
+            yData: Array<number> = [],
+            averageIndicator: any = this.averageIndicator,
+            isOHLC = isArray(yVal[0]);
+
+        // average indicator period minus 1
+        let range;
+
+        switch (averageType) {
+        case 'dema':
+            range = (2 * period - 1) - 1;
+            break;
+        case 'tema':
+            range = (3 * period - 2) - 1;
+            break;
+        case 'ema':
+        case 'wma':
+        default: // use sma if any of the above strings do not match
+            range = period - 1;
+        }
+
+        // Check period, if bigger than points length, skip
+        if (!index || yVal.length <= range) {
+            return;
+        }
+
+        // Get the average indicator's values
+        const values = averageIndicator.prototype
+            .getValues(series, params).yData;
+
+        // Get the Disparity Index indicator's values
+        for (let i = range; i < yValLen; i++) {
+            const disparityIndexValue: number = this.calculateDisparityIndex(
+                isOHLC ? (yVal[i] as any)[index] : yVal[i],
+                values[i - range]
+            );
+
+            disparityIndexPoint.push([
+                xVal[i],
+                disparityIndexValue
+            ]);
+            xData.push(xVal[i]);
+            yData.push(disparityIndexValue);
+        }
+
+        return {
+            values: disparityIndexPoint,
+            xData: xData,
+            yData: yData
+        } as IndicatorValuesObject<TLinkedSeries>;
+    }
+
+}
+
+interface DisparityIndexIndicator {
+    averageIndicator?:
+        typeof EMAIndicator|
+        typeof DEMAIndicator|
+        typeof SMAIndicator|
+        typeof TEMAIndicator|
+        typeof WMAIndicator;
+    pointClass: typeof DisparityIndexPoint;
+}
+
+extend(DisparityIndexIndicator.prototype, {
+    nameBase: 'Disparity Index'
```

Comment:
_Discusion:_
We should also show in the tooltip and legend the type of `average` we are using

Reply:
Good point!

Reply:
Yes, I thought about it. The problem I encountered is that I would have to define
```js
extend(DisparityIndexIndicator.prototype, {
    nameBase: some_name
});
```
in 2 different places. I send a commit with a code showing how would it look like now.

Reply:
Here is a commit: https://github.com/highcharts/highcharts/pull/15394/commits/24d8d6cf44cc13f341a6e12ebca9139261710ec0
As you can see, `nameBase` needs to be defined in 2 places:
- before indicator init, because it needs to be displayed in popup (otherwise it will be shown as `DISPARITYINDEX` instead of `Disparity Index`
![image](https://user-images.githubusercontent.com/23310742/113851069-65e28080-979b-11eb-812d-1364d02b387d.png)
https://github.com/highcharts/highcharts/blob/22fd2b1e8b995f5ce7cfad4ee96c5db4d3898dd6/ts/Extensions/Annotations/Popup.ts#L896-L905
- and while initing an indicator, when we know which average has been chosen.

Is it OK now?

Reply:
The average type should not be defined in the list in the popup. It's a parameter, so only tooltip/legend should include this info